src/tiff/man/tiffsplit.1
.\" $Id: tiffsplit.1,v 1.7 2016-09-25 20:05:52 bfriesen Exp $
.\"
.\" Copyright (c) 1992-1997 Sam Leffler
.\" Copyright (c) 1992-1997 Silicon Graphics, Inc.
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and 
.\" its documentation for any purpose is hereby granted without fee, provided
.\" that (i) the above copyright notices and this permission notice appear in
.\" all copies of the software and related documentation, and (ii) the names of
.\" Sam Leffler and Silicon Graphics may not be used in any advertising or
.\" publicity relating to the software without the specific, prior written
.\" permission of Sam Leffler and Silicon Graphics.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
.\" EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
.\" WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
.\" 
.\" IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
.\" ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
.\" OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
.\" LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
.\" OF THIS SOFTWARE.
.\"
.if n .po 0
.TH TIFFSPLIT 1 "September 20, 2005" "libtiff"
.SH NAME
tiffsplit \- split a multi-image
.SM TIFF
into single-image
.SM TIFF
files
.SH SYNOPSIS
.B tiffsplit
.I src.tif
[
.I prefix
]
.SH DESCRIPTION
.I tiffsplit
takes a multi-directory (page)
.SM TIFF
file and creates one or more single-directory (page)
.SM TIFF
files from it.
The output files are given names created by concatenating
a prefix, a lexically ordered
suffix in the range [\fIaaa\fP-\fIzzz\fP], the suffix
.I .tif 
(e.g. 
.IR xaaa.tif ,
.IR xaab.tif ,
.IR ... ,
.IR xzzz.tif ).
If a prefix is not specified on the command line,
the default prefix of
.I x
is used.
.SH OPTIONS
None.
.SH BUGS
Only a select set of ``known tags'' is copied when splitting.
.SH "SEE ALSO"
.BR tiffcp (1),
.BR tiffinfo (1),
.BR libtiff (3TIFF)
.PP
Libtiff library home page:
.BR http://www.simplesystems.org/libtiff/


debian/changelog
wxwidgets2.8 (2.8.1.1-0) unstable; urgency=low

  * Fix compilation with wxUSE_STL=1
  * wxGrid::GetBestSize() returns same size the grid would have after AutoSize()
  * Added wxTreeCtrl::CollapseAll[Children]() and IsEmpty() (Francesco Montorsi)
  * Several RTL-related positioning fixes (Diaa Sami)
  * Fix wxConfig::DeleteGroup() for arguments with trailing slash (David Hart)
  * Fix memory leak in wxGrid::ShowCellEditControl() (Christian Sturmlechner)
  * Don't crash if command line is not valid UTF-8 (Unicode build only)
  * Added wxSizerFlags::Shaped() and FixedMinSize() methods
  * Added wxCSConv::IsOk() (Manuel Martin)
  * Added wxDateTime::GetDateOnly()
  * Made wxTextFile work with unseekable files again (David Hart)
  * Added wxCONFIG_USE_SUBDIR flag to wxFileConfig (Giuseppe Bilotta)
  * Added wxSearchCtrl::[Get|Set]DescriptiveText

 -- Robin Dunn <robin@alldunn.com>  Fri, 19 Jan 2007 12:02:52 -0800

wxwidgets2.8 (2.8.0.1-1) unstable; urgency=low

  * No changes, just a different packager

 -- Vadim Zeitlin <vadim@wxwindows.org>  Dec, 13 Wed 2006 17:40:24 +0100

wxwidgets2.8 (2.8.0.1-0) unstable; urgency=low

  * Add wxSearchCtrl
  * Official wxPython release

 -- Robin Dunn <robin@alldunn.com>  Mon, 11 Dec 2006 18:50:29 -0800

wxwidgets2.8 (2.8.0.0-0rc3) unstable; urgency=low

  * wxFileSystem::OpenFile() takes wxFS_SEEKABLE by default now.
  * Fixed wxCalendarCtrl::HitTest() when clicking on month change arrows.
  * Added wxWindow::GetWindowBorderSize(), corrected wxTreeCtrl::GetBestSize().
  * Allow dynamically changing most of text control styles.
  * Enable use of libgnomeprintui by default in configure.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Dec, 05 Tue 2006 14:41:22 +0100

wxwidgets2.8 (2.8.0.0-0rc2) unstable; urgency=low

  * wxArrayString::Alloc() doesn't clear array contents any more
  * Fixed long standing bug in wxFileConfig groups renaming

 -- Vadim Zeitlin <vadim@wxwindows.org>  Nov, 14 Tue 2006 19:06:37 +0100

wxwidgets2.7 (2.7.2.0-0) unstable; urgency=low

  * Added wxTar streams.
  * Added wxFilterFSHandler and wxArchiveFSHandler.
  * Added wxString::ToLongLong() and ToULongLong().
  * Added wxCollapsiblePane.
  * Added wxSimpleHtmlListBox.
  * Many enhancements to wxRichTextCtrl.
  * Support for loading TGA files added.
  * wxHtmlWindow now generates events for mouse actions and text copy.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Oct, 31 Tue 2006 15:58:00 +0100

wxwidgets2.7 (2.7.1.3-0) unstable; urgency=low

  * wxGTK:  The wx.ALWAYS_SHOW_SB style is now supported.
  * Fixed name errors in the old wxPython package namespace.

 -- Robin Dunn <robin@alldunn.com>  Thu, 26 Oct 2006 14:35:29 -0700

wxwidgets2.7 (2.7.1.2-0) unstable; urgency=low

  * Fixed a bug in the MaskedEdit controls caused by conflicting IsEmpty methods.
  * Patch #1579280: Some mimetype optimizations on unix-like systems.

 -- Robin Dunn <robin@alldunn.com>  Sat, 21 Oct 2006 22:33:41 -0700

wxwidgets2.7 (2.7.1.1-0) unstable; urgency=low

  * First 2.7.x release from wxPython team.
  * Many enhancements, changes, fixes and updates.  See
    http://wxpython.org/CHANGES.txt for details.

 -- Robin Dunn <robin@alldunn.com>  Wed, 18 Oct 2006 17:09:53 -0700

wxwidgets2.7 (2.7.0.0-2) unstable; urgency=low

  * Fixes to ansi packages builds
  * Corresponds to the official wxWidgets 2.7.0 sources

 -- Vadim Zeitlin <vadim@wxwindows.org>  Fri,  6 Aug 2006 16:31:08 +0200

wxwidgets2.7 (2.7.0.0-1) unstable; urgency=low

  * New 2.7.0 release of the new development series.
  * It's now possible to build both ansi and unicode versions.
  * Maintainer change.

 -- Vadim Zeitlin <vadim@wxwindows.org>  Fri,  4 Aug 2006 16:12:47 +0200

wxwidgets2.7 (2.7.0.0-0rpd) unstable; urgency=low

  * First test build with wx 2.7 tree.

 -- Robin Dunn <robin@alldunn.com>  Thu,  1 Jun 2006 14:23:16 -0700

wxwidgets2.6 (2.6.3.2-3rpd) unstable; urgency=low

  * Added icons and *.desktop files for the wpython-wxtools

 -- Robin Dunn <robin@alldunn.com>  Thu,  4 May 2006 15:35:08 -0700

wxwidgets2.6 (2.6.3.2) unstable; urgency=low

  * Next try for an upload candidate.
  * Fixes FTBFS on GNU/kFreeBSD. Closes: #327638
  * Update fr.po, thanks to Thomas Huriaux and debian-l10n-french.
    Closes: #328966
  * Split wxbase back into its own package, after all these years
    and just after I drop it, someone finally has a use for it.
    Such is life.  Closes: #333944
  * Fixes FTBFS with new make.  If this ugly quoting is the tradeoff,
    I'm itching to learn of the fabulous benefits...
    Ack NMU for this from Kurt Roeckx to keep the amd64 transition
    rolling while my x86 build machine was painfully dying.  Thanks!
    Closes: #350695, #361443
  * New FSF address in copyright file.
  * Add an index.html link to the doc contents page.  Closes: #328896
  * Fixes declarations for friends, as required by C++, and now gcc4.1.
    Closes: #358219
  * Add an unpack_examples script, loosely based on one submitted
    by Richard Esplin.  Closes: #342761
  * Disable the reserved vtable entries (again, but this time its
    optional) to retain compatibility with the existing release
    in the distro.
  * Update build-deps for xorg7.

 -- Ron Lee <ron@debian.org>  Fri, 14 Apr 2006 16:50:24 +0930

wxwidgets2.6 (2.6.1.2-0.1) unstable; urgency=low

  * Non-maintainer upload.
  * Make it build using the new make.  Patch from
    Daniel Schepler <schepler@math.berkeley.edu> (Closes: #350695)

 -- Kurt Roeckx <kurt@roeckx.be>  Sun,  9 Apr 2006 14:04:34 +0200

wxwidgets2.6 (2.6.1.2) unstable; urgency=low

  * Fix the build_options.py created by the in tree build.
    Closes: #324960  Thanks.
  * Move the wx.pth alternative to its own directory where
    it can't mess with anything else.
  * Pull in changes to dcclient.cpp and window.cpp from HEAD
    mostly for gtk2.8 compatibility, but fixes a couple of
    other issues too.

 -- Ron Lee <ron@debian.org>  Thu, 25 Aug 2005 18:38:31 +0930

wxwidgets2.6 (2.6.1.1.1) unstable; urgency=low

src/tiff/tools/tiffsplit.c
/* $Id: tiffsplit.c,v 1.23 2015-05-28 13:10:26 bfriesen Exp $ */

/*
 * Copyright (c) 1992-1997 Sam Leffler
 * Copyright (c) 1992-1997 Silicon Graphics, Inc.
 *
 * Permission to use, copy, modify, distribute, and sell this software and 
 * its documentation for any purpose is hereby granted without fee, provided
 * that (i) the above copyright notices and this permission notice appear in
 * all copies of the software and related documentation, and (ii) the names of
 * Sam Leffler and Silicon Graphics may not be used in any advertising or
 * publicity relating to the software without the specific, prior written
 * permission of Sam Leffler and Silicon Graphics.
 * 
 * THE SOFTWARE IS PROVIDED "AS-IS" AND WITHOUT WARRANTY OF ANY KIND, 
 * EXPRESS, IMPLIED OR OTHERWISE, INCLUDING WITHOUT LIMITATION, ANY 
 * WARRANTY OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  
 * 
 * IN NO EVENT SHALL SAM LEFFLER OR SILICON GRAPHICS BE LIABLE FOR
 * ANY SPECIAL, INCIDENTAL, INDIRECT OR CONSEQUENTIAL DAMAGES OF ANY KIND,
 * OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER OR NOT ADVISED OF THE POSSIBILITY OF DAMAGE, AND ON ANY THEORY OF 
 * LIABILITY, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE 
 * OF THIS SOFTWARE.
 */

#include "tif_config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tiffio.h"

#ifndef HAVE_GETOPT
extern int getopt(int, char**, char*);
#endif

#define	CopyField(tag, v) \
    if (TIFFGetField(in, tag, &v)) TIFFSetField(out, tag, v)
#define	CopyField2(tag, v1, v2) \
    if (TIFFGetField(in, tag, &v1, &v2)) TIFFSetField(out, tag, v1, v2)
#define	CopyField3(tag, v1, v2, v3) \
    if (TIFFGetField(in, tag, &v1, &v2, &v3)) TIFFSetField(out, tag, v1, v2, v3)

#define PATH_LENGTH 8192

static const char TIFF_SUFFIX[] = ".tif";

static	char fname[PATH_LENGTH];

static	int tiffcp(TIFF*, TIFF*);
static	void newfilename(void);
static	int cpStrips(TIFF*, TIFF*);
static	int cpTiles(TIFF*, TIFF*);

int
main(int argc, char* argv[])
{
	TIFF *in, *out;

	if (argc < 2) {
                fprintf(stderr, "%s\n\n", TIFFGetVersion());
		fprintf(stderr, "usage: tiffsplit input.tif [prefix]\n");
		return (-3);
	}
	if (argc > 2) {
		strncpy(fname, argv[2], sizeof(fname));
		fname[sizeof(fname) - 1] = '\0';
	}
	in = TIFFOpen(argv[1], "r");
	if (in != NULL) {
		do {
			size_t path_len;
			char *path;
			
			newfilename();

			path_len = strlen(fname) + sizeof(TIFF_SUFFIX);
			path = (char *) _TIFFmalloc(path_len);
			strncpy(path, fname, path_len);
			path[path_len - 1] = '\0';
			strncat(path, TIFF_SUFFIX, path_len - strlen(path) - 1);
			out = TIFFOpen(path, TIFFIsBigEndian(in)?"wb":"wl");
			_TIFFfree(path);

			if (out == NULL)
				return (-2);
			if (!tiffcp(in, out))
				return (-1);
			TIFFClose(out);
		} while (TIFFReadDirectory(in));
		(void) TIFFClose(in);
	}
	return (0);
}

static void
newfilename(void)
{
	static int first = 1;
	static long lastTurn;
	static long fnum;
	static short defname;
	static char *fpnt;

	if (first) {
		if (fname[0]) {
			fpnt = fname + strlen(fname);
			defname = 0;
		} else {
			fname[0] = 'x';
			fpnt = fname + 1;
			defname = 1;
		}
		first = 0;
	}
#define	MAXFILES	17576
	if (fnum == MAXFILES) {
		if (!defname || fname[0] == 'z') {
			fprintf(stderr, "tiffsplit: too many files.\n");
			exit(1);
		}
		fname[0]++;
		fnum = 0;
	}
	if (fnum % 676 == 0) {
		if (fnum != 0) {
			/*
                         * advance to next letter every 676 pages
			 * condition for 'z'++ will be covered above
                         */
			fpnt[0]++;
		} else {
			/*
                         * set to 'a' if we are on the very first file
                         */
			fpnt[0] = 'a';
		}
		/*
                 * set the value of the last turning point
                 */
		lastTurn = fnum;
	}
	/* 
         * start from 0 every 676 times (provided by lastTurn)
         * this keeps us within a-z boundaries
         */
	fpnt[1] = (char)((fnum - lastTurn) / 26) + 'a';
	/* 
         * cycle last letter every file, from a-z, then repeat
         */
	fpnt[2] = (char)(fnum % 26) + 'a';
	fnum++;
}

static int
tiffcp(TIFF* in, TIFF* out)
{
	uint16 bitspersample, samplesperpixel, compression, shortv, *shortav;
	uint32 w, l;
	float floatv;
	char *stringv;
	uint32 longv;

	CopyField(TIFFTAG_SUBFILETYPE, longv);
	CopyField(TIFFTAG_TILEWIDTH, w);
	CopyField(TIFFTAG_TILELENGTH, l);
	CopyField(TIFFTAG_IMAGEWIDTH, w);
	CopyField(TIFFTAG_IMAGELENGTH, l);
	CopyField(TIFFTAG_BITSPERSAMPLE, bitspersample);
	CopyField(TIFFTAG_SAMPLESPERPIXEL, samplesperpixel);
	CopyField(TIFFTAG_COMPRESSION, compression);
	if (compression == COMPRESSION_JPEG) {
		uint32 count = 0;
		void *table = NULL;
		if (TIFFGetField(in, TIFFTAG_JPEGTABLES, &count, &table)
		    && count > 0 && table) {
		    TIFFSetField(out, TIFFTAG_JPEGTABLES, count, table);
		}
	}
        CopyField(TIFFTAG_PHOTOMETRIC, shortv);
	CopyField(TIFFTAG_PREDICTOR, shortv);
	CopyField(TIFFTAG_THRESHHOLDING, shortv);
	CopyField(TIFFTAG_FILLORDER, shortv);
	CopyField(TIFFTAG_ORIENTATION, shortv);
	CopyField(TIFFTAG_MINSAMPLEVALUE, shortv);
	CopyField(TIFFTAG_MAXSAMPLEVALUE, shortv);
	CopyField(TIFFTAG_XRESOLUTION, floatv);
	CopyField(TIFFTAG_YRESOLUTION, floatv);
	CopyField(TIFFTAG_GROUP3OPTIONS, longv);
	CopyField(TIFFTAG_GROUP4OPTIONS, longv);
	CopyField(TIFFTAG_RESOLUTIONUNIT, shortv);
	CopyField(TIFFTAG_PLANARCONFIG, shortv);
	CopyField(TIFFTAG_ROWSPERSTRIP, longv);
	CopyField(TIFFTAG_XPOSITION, floatv);
	CopyField(TIFFTAG_YPOSITION, floatv);
	CopyField(TIFFTAG_IMAGEDEPTH, longv);
	CopyField(TIFFTAG_TILEDEPTH, longv);
	CopyField(TIFFTAG_SAMPLEFORMAT, shortv);
	CopyField2(TIFFTAG_EXTRASAMPLES, shortv, shortav);
	{ uint16 *red, *green, *blue;
	  CopyField3(TIFFTAG_COLORMAP, red, green, blue);
	}
	{ uint16 shortv2;
	  CopyField2(TIFFTAG_PAGENUMBER, shortv, shortv2);
	}
	CopyField(TIFFTAG_ARTIST, stringv);
	CopyField(TIFFTAG_IMAGEDESCRIPTION, stringv);
	CopyField(TIFFTAG_MAKE, stringv);
	CopyField(TIFFTAG_MODEL, stringv);
	CopyField(TIFFTAG_SOFTWARE, stringv);
	CopyField(TIFFTAG_DATETIME, stringv);
	CopyField(TIFFTAG_HOSTCOMPUTER, stringv);
	CopyField(TIFFTAG_PAGENAME, stringv);
	CopyField(TIFFTAG_DOCUMENTNAME, stringv);
	CopyField(TIFFTAG_BADFAXLINES, longv);
	CopyField(TIFFTAG_CLEANFAXDATA, longv);
	CopyField(TIFFTAG_CONSECUTIVEBADFAXLINES, longv);
	CopyField(TIFFTAG_FAXRECVPARAMS, longv);
	CopyField(TIFFTAG_FAXRECVTIME, longv);
	CopyField(TIFFTAG_FAXSUBADDRESS, stringv);
	CopyField(TIFFTAG_FAXDCS, stringv);
	if (TIFFIsTiled(in))
		return (cpTiles(in, out));
	else
		return (cpStrips(in, out));
}

static int
cpStrips(TIFF* in, TIFF* out)
{
	tmsize_t bufsize  = TIFFStripSize(in);
	unsigned char *buf = (unsigned char *)_TIFFmalloc(bufsize);

	if (buf) {
		tstrip_t s, ns = TIFFNumberOfStrips(in);
		uint64 *bytecounts;

		if (!TIFFGetField(in, TIFFTAG_STRIPBYTECOUNTS, &bytecounts)) {
			fprintf(stderr, "tiffsplit: strip byte counts are missing\n");
                        _TIFFfree(buf);
			return (0);
		}
		for (s = 0; s < ns; s++) {
			if (bytecounts[s] > (uint64)bufsize) {
				buf = (unsigned char *)_TIFFrealloc(buf, (tmsize_t)bytecounts[s]);
				if (!buf)
					return (0);
				bufsize = (tmsize_t)bytecounts[s];
			}
			if (TIFFReadRawStrip(in, s, buf, (tmsize_t)bytecounts[s]) < 0 ||
			    TIFFWriteRawStrip(out, s, buf, (tmsize_t)bytecounts[s]) < 0) {
				_TIFFfree(buf);
				return (0);
			}
		}
		_TIFFfree(buf);
		return (1);
	}
	return (0);
}

static int
cpTiles(TIFF* in, TIFF* out)
{
	tmsize_t bufsize = TIFFTileSize(in);
	unsigned char *buf = (unsigned char *)_TIFFmalloc(bufsize);

	if (buf) {
		ttile_t t, nt = TIFFNumberOfTiles(in);
		uint64 *bytecounts;

		if (!TIFFGetField(in, TIFFTAG_TILEBYTECOUNTS, &bytecounts)) {
			fprintf(stderr, "tiffsplit: tile byte counts are missing\n");
                        _TIFFfree(buf);
			return (0);
		}
		for (t = 0; t < nt; t++) {
			if (bytecounts[t] > (uint64) bufsize) {
				buf = (unsigned char *)_TIFFrealloc(buf, (tmsize_t)bytecounts[t]);
				if (!buf)
					return (0);
				bufsize = (tmsize_t)bytecounts[t];
			}
			if (TIFFReadRawTile(in, t, buf, (tmsize_t)bytecounts[t]) < 0 ||
			    TIFFWriteRawTile(out, t, buf, (tmsize_t)bytecounts[t]) < 0) {
				_TIFFfree(buf);
				return (0);
			}
		}
		_TIFFfree(buf);
		return (1);
	}
	return (0);
}

/* vim: set ts=8 sts=8 sw=8 noet: */
/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 8
 * fill-column: 78
 * End:
 */
